bundle/edu.gemini.pot/src/main/java/edu/gemini/spModel/target/obsComp/TargetSelection.java

```diff
@@ -34,110 +36,144 @@ private TargetSelection() {}
+
+            int idx = 0;
+            final List<Selection> res = new ArrayList<>();
+
+            res.add(new Selection(idx++, None.instance(), new Some<>(env.getBase())));
+            for (final GuideGroup g : env.getGroups()) {
+                final Option<GuideGroup> gOpt = new Some<>(g);
+                res.add(new Selection(idx++, gOpt, None.instance()));
+                for (final Option<SPTarget> tOpt : g.getTargets().map(Some::new)) {
+                    res.add(new Selection(idx++, gOpt, tOpt));
+                }
+            }
+            for (final Option<SPTarget> tOpt : env.getUserTargets().map(Some::new)) {
+                res.add(new Selection(idx++, None.instance(), tOpt));
+            }
+            return DefaultImList.create(res);
+        }
     }
 
-    public static int indexOf(final TargetEnvironment env, final GuideGroup grp) {
-        return toSelections(env).zipWithIndex().
-                find(tup -> grp.equals(tup._1().guideGroup)).
-                map(INDEX_OF).getOrElse(NO_SELECTION.value);
+    public static Option<Integer> indexOfTarget(final TargetEnvironment env, final SPTarget target) {
+        return Selection.toSelections(env).find(sel -> sel.target.exists(target::equals)).map(sel -> sel.index);
     }
 
-    private static Selection selectionAt(final TargetEnvironment env, final int index) {
-        if (index < 0) return null;
-        final ImList<Selection> lst = toSelections(env);
-        return (index < lst.size()) ? lst.get(index) : null;
-    }
+    public static Option<Integer> indexOfIndexedGuideGroup(final TargetEnvironment env,
+                                                           final IndexedGuideGroup igg) {
+        // Extract the selections (and their indices) corresponding to guide groups.
+        final ImList<Selection> filtered = Selection.toSelections(env).filter(Selection::isGuideGroup);
 
-    public static SPTarget get(final TargetEnvironment env, final int index) {
-        final Selection s = selectionAt(env, index);
-        return (s == null) ? null : s.target;
+        // Now make sure the guide group at position gpIdx is gp, and if so, return its index in the selections.
+        final int gpIdx     = igg.index();
+        final GuideGroup gp = igg.group();
+        if (gpIdx >= 0 && gpIdx < filtered.size()) {
+            final Selection sel = filtered.get(gpIdx);
+            return sel.guideGroup.filter(gp::equals).map(ign -> sel.index);
+        } else {
+            return None.instance();
+        }
     }
 
-    public static SPTarget get(final TargetEnvironment env, final ISPNode node) {
-        return get(env, getIndex(node));
+    private static Option<Selection> selectionAtIndex(final TargetEnvironment env, final int index) {
+        if (index < 0) return None.instance();
+        final ImList<Selection> lst = Selection.toSelections(env);
+        return ImOption.apply(index < lst.size() ? lst.get(index) : null);
```

Comment:
We should probably add a `getOption` method to `ImList` which does the range check for you.